### ISSUE TITLE
Feature: devtool/build - Podman support

### ISSUE THREAD

**Issue by bbros-dev** [firecracker-microvm/firecracker#1973]
We don't have Docker. We likely won't ever.  Docker only support can be worked around, but it would be nicer to have `devtool/build` not care.

In our experience so far the tools are interchangeable so there should be no additional maintenance burden.

Would adding Podman support to `devtool/build` be acceptable?

**Comment by tnotw** [comment]
Are there any cgroups version considerations to doing this?  For example, the Fedora (post-31) defaults are podman/cgroups v2.  For me, podman worked ok, but the tests for the jailer failed (I believe it still requires cgroups v1), so I had to revert to docker/cgroups v1.

**Comment by bbros-dev** [comment]
TL;DR: No not for these changes. I think ;)

> For me, podman worked ok

Interesting? Was that as parameterized and configured in `devtool`, or using your own incantation?

As rootless it should have broken on the `--tty` with the device trees shared.

My understanding is the cgroup v1 vs v2 is about what engine you can run, or vice versa, the engine you run (currently) sets the cgroup ver you use.

When we turn to determining the actual/minimal capabilities each contain is granted then the v1 vs v2 issue will loom. This change doesn't address that.

Frankyl - I hope that shortly we can dump these containers and have Firecracker uVM build Firecracker.
In fact I'd suggest the time investigating the "least permissions required" for these containers could be better spent on porting the build to a couple of uVM's.

Or is there some benefit from the container setup that I am missing?

**Comment by tnotw** [comment]
It was my own incantation.  It was probably just sufficient to run, but nothing more. 

This is the error that led me to believe rootless podman (which appears to require cgroups v2) might not fly: 
```
E   Exception: thread 'main' panicked at 'Jailer error: cpu configurations not found in /proc/mounts', src/jailer/src/main.rs:353:27
```
Which I interpreted as the the jailer confirming the cgroups v1 structures in /proc/mounts,  which don't appear to be there under Fedora/cgroups v2.  Perhaps this is just a Fedora problem,  or some other cgroups configuration issue though.

**Comment by bbros-dev** [comment]
Thanks for the additional info. +1 for uVM builds.

**Comment by acatangiu** [comment]
> I hope that shortly we can dump these containers and have Firecracker uVM build Firecracker.
Or is there some benefit from the container setup that I am missing?

One problem to using uVM builds is the need for virtualization support in dev machines.
Any dev environment running on a VM without nested virtualization support would be unusable.

Containers are a familiar and accepted way for deploying a build environment (our particular case) and they come with services and tools that make building and distribution easy. I don't personally see value in opening a uVM images store/hub :) for example.

Having said that, I do agree that we can _improve_ our container setup.

**Comment by bbros-dev** [comment]
The issue of running the fcuvm/dev container as root with elevated privileges is thorny. We likely can never run (at least) some parts of the test suite without elevated privileges, doing all the testing and analysis to work out which test items require which capabilities is a substantial effort and (in our opinion) likely not the best use of dev cycles right now.
Example: a higher priority is writing, documenting and working out the equivalent jailer details.

There are two workarounds readily available to anyone who comes along this path and wants to run the test suite, in order of ease:

Workaround 1:
Only run the test suite on external CI/CD providers. These generally run as root, and your systems aren't touched by any libraries the test suite pulls in and runs.

Workaround 2:
Only run the test suite as rootless, either Docker (hard to configure) or Podman (no setup required).
The Podman path is provided by PR #1973 and the instructions therein.

Closing as fixed for us, and for anyone else there are two workarounds at hand.
